### PR TITLE
[CIVIS-6117] ruby 3.1, 3.2, and 3.3 upgrade

### PR DESCRIPTION
**Pull request by thatguyinabeanie**
## Description

* Makes authority gem code compatible with ruby 3.1, 3.2, and 3.3
* tested against all major versions of rails. 

## Context, Consequences, & Considerations
Required: Please step through the following list, pausing at each item to consider your change in relation to the item's context.
Check the box to mark that it applies, and enter your relevant notes under the item.

- [ ] Security: This has security implications. This includes (but not limited to) adding users, modifying user/app permissions, network rules/policies, changing a system interconnection, or changing an authorization strategy.
  - [x] This PR does not require security review. These changes are part of a project plan that has already undergone security review. The link is provided below.
  - [ ] This PR requires security review. Add the `security` label to this PR then request a review from the [Security Code Reviewers Team](https://github.com/orgs/civisanalytics/teams/security-code-reviewers).

>

- [ ] Execution: This change requires commands to be run outside of the normal merge.

>

- [ ] Execution: Applying this change will modify production data. This includes (but not limited to) database migrations, rake tasks that modify data, or other processes.

>

- [ ] Impact: This change may cause service interruptions.

>

- [ ] Testing: How did you test this change (unit tests, acceptance tests, etc.)? Did you do any manual testing?

>

- [ ] Testing: How will you confirm this change once it's merged?

>

- [ ] Documentation: Documentation to reflect this change has been added to Confluence or Zendesk.

>

- [x] **All items of the checklist have been considered and this PR description is complete.**

### CODE REVIEW

**Comment by thatguyinabeanie** [comment]
This project does not currently support gating on a green travis build.
i already reached out to @Naphier to either get that setup or be given admin permissions on the repo so I can set it up. either way is fine.

for now i'll just have to link the most recent travis build.

**Comment by thatguyinabeanie** [comment]
image

https://app.travis-ci.com/github/civisanalytics/authority/builds/270466561?serverType=git

File: .travis.yml [hunk 1]
```diff
@@ -1,11 +1,18 @@
 language: ruby
 rvm:
-  - 1.9.3
-  - 2.0.0
-  - 2.1.10
-  - 2.2.6
-  - 2.3.3
-  - 2.4.0
+  # - 1.9.3
+  # - 2.0.0-p648
```
Comment (thatguyinabeanie) [inline .travis.yml]: i guess this means dropping support for ruby 1.x and 2.0.x /shrug
v 1.9.3 and 2.0.0 gave me problems. In travis, rvm install couldn't find these versions for some reason. I didn't try particular long to try to resolve that. given that they're very old versions, i just commented it out and went on my happy way

i guess this really only matters if we think we're going to want to merge this back into the main repo.

File: .travis.yml [hunk 2]
```diff
@@ -17,21 +24,160 @@ gemfile:
   - gemfiles/4.0.gemfile
   - gemfiles/4.1.gemfile
   - gemfiles/4.2.gemfile
+  - gemfiles/5.0.gemfile
+  - gemfiles/5.1.gemfile
+  - gemfiles/5.2.gemfile
+  - gemfiles/6.0.gemfile
+  - gemfiles/6.1.gemfile
+  - gemfiles/7.0.gemfile
+  - gemfiles/7.1.gemfile
```
Comment (thatguyinabeanie) [inline .travis.yml:37]: i added 7.0 and 7.1 after fixing 6.1. and to my surprise it all worked without additional changes

File: .travis.yml [hunk 3]
```diff
@@ -17,21 +24,160 @@ gemfile:
...
+
+script: bundle exec rspec
 
 matrix:
   exclude:
-    - rvm: rbx
-      gemfile: gemfiles/4.0.gemfile
-    - rvm: 2.4.0
+    # Rails 3.2
```
Comment (thatguyinabeanie) [inline .travis.yml:43]: manually ignoring ruby,rails version combinations that are not compatible.

File: .travis.yml (hunk 1 above)
Comment (leanne73) [reply .travis.yml]: Huh, yeah, I think it's reasonable to drop support for those versions now.

The original repo has been archived, so we don't have to worry about merging this back in.

File: .travis.yml [hunk 4]
```diff
@@ -17,21 +24,160 @@ gemfile:
...
+      gemfile: gemfiles/6.1.gemfile
+    - rvm: 2.1
+      gemfile: gemfiles/6.1.gemfile
+    - rvm: 2.2
+      gemfile: gemfiles/6.1.gemfile
+    - rvm: 2.3
+      gemfile: gemfiles/6.1.gemfile
+    - rvm: 2.4
+      gemfile: gemfiles/6.1.gemfile
+    - rvm: 3.0
```
Comment (leanne73) [inline .travis.yml]: I thought that 3.0 was the [recommended](https://www.fastruby.io/blog/ruby/rails/versions/compatibility-table.html) version of ruby for rails 6.1? And compatible with 7 & 7.1 too. Unless I'm reading this list wrong. But it looks like 3.0 is excluded for every version of Rails - seems like it should be compatible with a couple versions at least.

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -1,6 +1,8 @@
 $LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
 require 'rspec'
 require 'authority'
+require 'pry-byebug'
+require 'pry'
```
Comment (leanne73) [inline spec/spec_helper.rb]: Nit: [`pry-byebug`](https://github.com/deivid-rodriguez/pry-byebug/blob/master/pry-byebug.gemspec#L25) should take care of installing `pry` for us

File: .travis.yml (hunk 4 above)
Comment (thatguyinabeanie) [reply .travis.yml]: thanks for catching that. i had some issues with the build not being able to install 3.0 early on. I overlooked going back to revisit this. working on it now

Comment (thatguyinabeanie) [reply .travis.yml]: fixed it!

**Comment by thatguyinabeanie** [comment]
latest travis ci build
https://app.travis-ci.com/github/civisanalytics/authority/builds/270483337?serverType=git

image

**Comment by thatguyinabeanie** [comment]
image

done done for real now

https://app.travis-ci.com/github/civisanalytics/authority/builds/270507711?serverType=git

**Comment by thatguyinabeanie** [comment]
adding @gskifstad and @aehrmann since Leanne is going off-call 2maro on friday.

**Comment by thatguyinabeanie** [comment]
https://app.travis-ci.com/github/civisanalytics/authority/builds/270513183?serverType=git

done done for real for real

**Review by aehrmann — DISMISSED** [review]
LGTM

File: README.markdown [hunk 6]
```diff
@@ -15,38 +7,39 @@ Authority will work fine with a standalone app or a single sign-on system. You c
 If you're using it with Rails controllers, it requires that you already have some kind of user object in your application, accessible via a method like `current_user` (configurable).
 
 [![Gem Version](https://badge.fury.io/rb/authority.png)](https://rubygems.org/gems/authority)
```
Comment (leanne73) [inline README.markdown:9]: Would be good to update this when we know where we'll be hosting our version of the gem.

File: spec/authority/configuration_spec.rb [hunk 7]
```diff
@@ -15,7 +15,8 @@
 
       it "logs to standard error by default" do
         Authority.instance_variable_set :@configuration, nil
-        null = File.exists?('/dev/null') ? '/dev/null' : 'NUL:' # Allow for Windows
+        method = RUBY_VERSION < '3.2.0' ? :exists? : :exist?
```
Comment (leanne73) [inline spec/authority/configuration_spec.rb]: I forgot to mention this earlier, but we should safely be able to use `exist?` with all versions of ruby. `exists?` was just an alias for it!

https://ruby-doc.org/core-2.1.0/File.html#exist-3F-method

Comment (thatguyinabeanie) [reply spec/authority/configuration_spec.rb]: this was true until the alias was removed in 3.2.

Comment (leanne73) [reply spec/authority/configuration_spec.rb]: Right, we _need_ to use `exist?` in 3.0+, but we can also use it in earlier versions! So we shouldn't need the conditional, we can just use `exist?`. link above is to the 2.1 docs which show that the `exist?` method was already in existence then (lol), and that's the earliest ruby version that the gem still supports

**Review by aehrmann — DISMISSED** [review]

**Review by leanne73 — COMMENTED** [review]
Looks like there are still some Travis errors, but thanks for fixing the nit about the exists method and updating the gem name.

I think we can update the readme with the new rubygems link later, once we've pushed the first version up!

**Comment by thatguyinabeanie** [comment]
okay. finally good to go. https://app.travis-ci.com/github/civisanalytics/authority/builds/270551328?serverType=git
image

File: lib/authority/controller.rb [hunk 8]
```diff
@@ -111,7 +109,7 @@ def add_actions(action_map)
       # for all Rails actions in the action map
       def force_action(forced_action)
         add_actions(
-          authority_action_map.transform_values {|_| forced_action }
```
Comment (thatguyinabeanie) [inline lib/authority/controller.rb]: when i added rubocop and ran it locally with auto-fix, rubocop changed the code to this line, which is not compatible with older versions of ruby.

undid this change and removed rubocop for the time being. we can revist adding that later.

**Review by aehrmann — APPROVED** [review]
LGTM

**Pull request merged**